polybench/correlation.cpp

```diff
@@ -0,0 +1,218 @@
+	}
+
+	// Center and reduce the column vectors.
+	for(size_t i = 1; i <= N; i++) {
+		for(size_t j = 1; j <= M; j++) {
+			data[i * (M + 1) + j] -= mean[j];
+			data[i * (M + 1) + j] /= sqrt(FLOAT_N);
+			data[i * (M + 1) + j] /= stddev[j];
+		}
+	}
+
+	// Calculate the m * m correlation matrix.
+	for(size_t j1 = 1; j1 <= M - 1; j1++) {
+		symmat[j1 * (M + 1) + j1] = 1.0;
+
+		for(size_t j2 = j1 + 1; j2 <= M; j2++) {
+			symmat[j1 * (M + 1) + j2] = 0.0;
+
+			for(size_t i = 1; i <= N; i++) {
+				symmat[j1 * (M + 1) + j2] += (data[i * (M + 1) + j1] * data[i * (M + 1) + j2]);
+			}
+
+			symmat[j2 * (M + 1) + j1] = symmat[j1 * (M + 1) + j2];
+		}
+	}
+
+	symmat[M * (M + 1) + M] = 1.0;
+}
+
+class Polybench_Correlation {
+  public:
+	Polybench_Correlation(const BenchmarkArgs& args) : args(args), size(args.problem_size) {}
+
+	void setup() {
+		data.resize((size + 1) * (size + 1));
+		mean.resize(size + 1);
+		stddev.resize(size + 1);
+		symmat.resize((size + 1) * (size + 1));
+
+		init_arrays(data.data(), size);
+	}
+
+	void run() {
+		using namespace cl::sycl;
+
+		buffer<DATA_TYPE, 2> data_buffer{data.data(), range<2>(size + 1, size + 1)};
+		buffer<DATA_TYPE, 1> mean_buffer{mean.data(), range<1>(size + 1)};
+		buffer<DATA_TYPE, 1> stddev_buffer{stddev.data(), range<1>(size + 1)};
+		buffer<DATA_TYPE, 2> symmat_buffer{symmat.data(), range<2>(size + 1, size + 1)};
+
+		args.device_queue.submit([&](handler& cgh) {
+			auto data = data_buffer.get_access<access::mode::read>(cgh);
+			auto mean = mean_buffer.get_access<access::mode::read_write>(cgh);
+
+			cgh.parallel_for<class CorrelationMean>(range<1>(size), id<1>(1), [=, N_ = size](item<1> item) {
+				const auto j = item[0];
+
+				for(size_t i = 1; i <= N_; i++) {
+					mean[item] += data[{i, j}];
+				}
+				mean[item] /= ((DATA_TYPE)FLOAT_N);
+			});
+		});
+
+		args.device_queue.submit([&](handler& cgh) {
+			auto data = data_buffer.get_access<access::mode::read>(cgh);
+			auto mean = mean_buffer.get_access<access::mode::read>(cgh);
+			auto stddev = stddev_buffer.get_access<access::mode::read_write>(cgh);
+
+			cgh.parallel_for<class CorrelationStd>(range<1>(size), id<1>(1), [=, N_ = size](item<1> item) {
+				const auto j = item[0];
+
+				for(size_t i = 1; i <= N_; i++) {
+					stddev[item] += (data[{i, j}] - mean[item]) * (data[{i, j}] - mean[item]);
+				}
+
+				stddev[item] /= FLOAT_N;
+				stddev[item] = sqrt(stddev[item]);
+				stddev[item] = stddev[item] <= EPS ? 1.0 : stddev[item];
+			});
+		});
+
+		args.device_queue.submit([&](handler& cgh) {
+			auto data = data_buffer.get_access<access::mode::read_write>(cgh);
+			auto mean = mean_buffer.get_access<access::mode::read>(cgh);
+			auto stddev = stddev_buffer.get_access<access::mode::read>(cgh);
+
+			cgh.parallel_for<class CorrelationReduce>(range<2>(size, size), id<2>(1, 1), [=](item<2> item) {
+				const auto j = item[1];
+
+				data[item] -= mean[j];
+				data[item] /= sqrt(FLOAT_N);

```

Comment:
At least on hipSYCL CPU backend, this results in an ambiguous function call error because of collision between SYCL sqrt and `std::sqrt()`. This is solved by calling `sycl::sqrt()` instead.